semgrep-core/src/core/Report.ml

```diff
@@ -1,9 +1,79 @@
+ * not being used, we don't want to save them. On the other hand, we want
+ * to feel confident in the correctness of the code and make it easy to
+ * know what is or isn't being saved. And we want our code to be relatively
+ * readable.
+ *
+ * The debug_info type attempts to solve this. It has a variant for each
+ * verbosity mode semgrep-core may be invoked with, which contains all
+ * the fields semgrep requests from semgrep-core in that mode. Each result
+ * contains debug_info in addition to the always-reported fields. The
+ * variant of debug_info used in the results is always determined either
+ * by the mode, which is a global set in Main.ml after the arguments are
+ * read, or by a previous result. In this way we ensure that the fields
+ * stored in the result are determined by the arguments passed by the user.
+ *
+ * Alternatives considered to the extra field:
+ * - storing the information but just ommitting it in the final result (I
+ *   tried this but it still uses too much memory)
+ * - using option types within the result types for field and a global
+ *   config to decide which fields are in use
+ *
+ * I considered the latter, but I'm not a fan of mix-and-match option types.
+ * Maybe I just think it makes it tempting to use map_option, where here
+ * the assumption that is being made feels more obvious. It also makes it
+ * extra clear what information is being used for each mode and encourages
+ * us to do the work of deciding what should be included instead of giving
+ * the user choice that they probably don't know how to make intelligently.
+ * That being said, if we wanted users to be able to "mix-and-match" request
+ * fields, we should move to the option-types paradigm. Also, the collate
+ * functions are quite ugly. I'm open to argument.
+ *
+ * TODO: We could use atdgen to specify those to factorize type definitions
  * that have to be present anyway in Output_from_core.atd.
- *)
+ ********************************************************************************)
 
+let logger = Logging.get_logger [ __MODULE__ ]
+
+(********************************************************************************)
+(* Options for what extra debugging information to output.
+   These are generally memory intensive fields that aren't strictly needed *)
+(********************************************************************************)
+
+(* Coupling: the debug_info variant of each result record should always
+      be the same as the mode's variant *)
+type debug_mode = MDebug | MTime | MNo_info [@@deriving show]
+
+type 'a debug_info =
+  (* -debug: save all the information that could be useful *)
+  | Debug of {
+      skipped_targets : Output_from_core_t.skipped_target list;
+      profiling : 'a;
+    }
+  (* -json_time: save just profiling information; currently our metrics record this *)
+  | Time of 'a
+  (* save nothing else *)
+  | No_info
+[@@deriving show]
+
+let mode = ref MNo_info
```

Comment:
Don't we have the config available somewhere when we are going to build a `Report`?

Reply:
It would be pretty annoying (would have to pass the config into a lot of places). Would you prefer it if I moved it to Flag_semgrep so at least all our globals were in the same place?